### ISSUE TITLE
```#nullable enable``` causes unexpected compilation errors

### ISSUE THREAD

**Issue by AlekseyTs** [dotnet/roslyn#79098]
Compile the following code with ```#nullable enable```:
```
public class C1<T> where T : notnull
{
    public T F => throw null!;

    public static C1<T> operator &(C1<T> x, C1<T> y)
    {
        return x;
    }

    public static bool operator true(C1<T> x) => true;
    public static bool operator false(C1<T> x) => false;
}

public class C2
{}

class Program
{
    static void Main()
    {
        var x = Get((C2?)null);
        var y = Get(new C2());
        (x && y).F.ToString();
        (y && x).F.ToString();
        (y && y).F.ToString();
    }

    static C1<T> Get<T>(T x)
    {
        throw null!;
    }
}
```

Observed the following unexpected errors:
```
// /Program.cs(25,10): error CS0217: In order to be applicable as a short circuit operator a user-defined logical operator ('C1<C2>.operator &(C1<C2>, C1<C2>)') must have the same return type and parameter types
//         (x && y).F.ToString();
Diagnostic(ErrorCode.ERR_BadBoolOp, "x && y").WithArguments("C1<C2>.operator &(C1<C2>, C1<C2>)").WithLocation(25, 10),
// /Program.cs(26,10): error CS0217: In order to be applicable as a short circuit operator a user-defined logical operator ('C1<C2>.operator &(C1<C2>, C1<C2>)') must have the same return type and parameter types
//         (y && x).F.ToString();
Diagnostic(ErrorCode.ERR_BadBoolOp, "y && x").WithArguments("C1<C2>.operator &(C1<C2>, C1<C2>)").WithLocation(26, 10),
// /Program.cs(27,10): error CS0217: In order to be applicable as a short circuit operator a user-defined logical operator ('C1<C2>.operator &(C1<C2>, C1<C2>)') must have the same return type and parameter types
//         (y && y).F.ToString();
Diagnostic(ErrorCode.ERR_BadBoolOp, "y && y").WithArguments("C1<C2>.operator &(C1<C2>, C1<C2>)").WithLocation(27, 10)
```

The errors go away when compiled with ```#nullable disable```.

Expected: The errors should not be reported. The operator satisfy the requirements about which the errors are complaining.

**Comment by bernd5** [comment]
The issue disappears if you remove " where T : notnull" from `C1` or if you add that to `Get<T>`.

But AFAIK null-violations should only issue warnings - so it is a bug...
